File: packages/shared/src/types/shop.ts
/**
 * 정비소 관련 공통 타입 정의
 */

/**
 * 정비소 상태 열거형
 */
export enum ShopStatus {
  ACTIVE = 'active',
  INACTIVE = 'inactive',
  PENDING = 'pending',
  SUSPENDED = 'suspended'
}

/**
 * 정비소 타입 열거형
 */
export enum ShopType {
  GENERAL = 'general',           // 종합 정비소
  SPECIALIZED = 'specialized',   // 전문 정비소
  DEALER = 'dealer',             // 딜러 서비스 센터
  FRANCHISE = 'franchise',       // 프랜차이즈
  MOBILE = 'mobile',             // 이동식 정비소
  TIRE = 'tire',                 // 타이어 전문점
  BODY = 'body',                 // 바디 샵
  OIL = 'oil'                    // 오일 교환 전문점
}

/**
 * 정비소 서비스 유형 열거형
 */
export enum ServiceType {
  OIL_CHANGE = 'oil_change',             // 오일 교환
  TIRE_SERVICE = 'tire_service',         // 타이어 서비스
  BRAKE_SERVICE = 'brake_service',       // 브레이크 서비스
  ENGINE_REPAIR = 'engine_repair',       // 엔진 수리
  TRANSMISSION = 'transmission',         // 변속기 수리
  ELECTRICAL = 'electrical',             // 전기 시스템
  AC_SERVICE = 'ac_service',             // 에어컨 서비스
  BODY_REPAIR = 'body_repair',           // 바디 수리
  PAINTING = 'painting',                 // 도장
  INSPECTION = 'inspection',             // 정기 검사
  DIAGNOSTICS = 'diagnostics',           // 진단
  SUSPENSION = 'suspension',             // 서스펜션
  EXHAUST = 'exhaust',                   // 배기 시스템
  GLASS = 'glass',                       // 유리 교체/수리
  WHEEL_ALIGNMENT = 'wheel_alignment'    // 휠 얼라인먼트
}

/**
 * 정비소 인터페이스
 */
export interface Shop {
  id: string;
  name: string;
  type: ShopType;
  status: ShopStatus;
  description?: string;
  address: Address;
  contact: {
    phone: string;
    email?: string;
    website?: string;
  };
  businessHours: BusinessHours;
  services: ServiceType[];
  specialties?: string[];
  certifications?: string[];
  rating?: number;
  reviewCount?: number;
  images?: string[];
  coverImage?: string;
  location: {
    latitude: number;
    longitude: number;
  };
  staffCount?: number;
  bayCount?: number;
  established?: number;
  owner?: {
    id: string;
    name: string;
  };
  amenities?: string[];
  paymentMethods?: string[];
  brands?: string[];
  featured?: boolean;
  createdAt: string;
  updatedAt: string;
}

/**
 * 주소 인터페이스
 */
export interface Address {
  street: string;
  city: string;
  state: string;
  postalCode: string;
  country: string;
  formattedAddress?: string;
}

/**
 * 영업 시간 인터페이스
 */
export interface BusinessHours {
  monday: DayHours;
  tuesday: DayHours;
  wednesday: DayHours;
  thursday: DayHours;
  friday: DayHours;
  saturday: DayHours;
  sunday: DayHours;
  holidays?: HolidayHours[];
}

/**
 * 일일 영업 시간 인터페이스
 */
export interface DayHours {
  open: boolean;
  openTime?: string;  // HH:MM 형식
  closeTime?: string; // HH:MM 형식
  breakStart?: string; // HH:MM 형식
  breakEnd?: string;   // HH:MM 형식
}

/**
 * 공휴일 영업 시간 인터페이스
 */
export interface HolidayHours {
  date: string;      // YYYY-MM-DD 형식
  name: string;
  open: boolean;
  openTime?: string;  // HH:MM 형식
  closeTime?: string; // HH:MM 형식
}

/**
 * 정비소 리뷰 인터페이스
 */
export interface ShopReview {
  id: string;
  shopId: string;
  userId: string;
  userName: string;
  userAvatar?: string;
  rating: number;
  comment: string;
  serviceDate?: string;
  serviceType?: ServiceType[];
  vehicleInfo?: {
    make: string;
    model: string;
    year: number;
  };
  likes?: number;
  response?: {
    text: string;
    date: string;
    name: string;
  };
  images?: string[];
  createdAt: string;
  updatedAt: string;
}

/**
 * 정비소 생성 인터페이스
 */
export interface ShopCreate {
  name: string;
  type: ShopType;
  description?: string;
  address: Address;
  contact: {
    phone: string;
    email?: string;
    website?: string;
  };
  businessHours: BusinessHours;
  services: ServiceType[];
  specialties?: string[];
  certifications?: string[];
  images?: string[];
  coverImage?: string;
  location: {
    latitude: number;
    longitude: number;
  };
  staffCount?: number;
  bayCount?: number;
  established?: number;
  amenities?: string[];
  paymentMethods?: string[];
  brands?: string[];
}

/**
 * 정비소 업데이트 인터페이스
 */
export interface ShopUpdate {
  name?: string;
  type?: ShopType;
  status?: ShopStatus;
  description?: string;
  address?: Partial<Address>;
  contact?: Partial<{
    phone: string;
    email?: string;
    website?: string;
  }>;
  businessHours?: Partial<BusinessHours>;
  services?: ServiceType[];
  specialties?: string[];
  certifications?: string[];
  images?: string[];
  coverImage?: string;
  location?: {
    latitude: number;
    longitude: number;
  };
  staffCount?: number;
  bayCount?: number;
  established?: number;
  amenities?: string[];
  paymentMethods?: string[];
  brands?: string[];
  featured?: boolean;
}

/**
 * 정비소 필터 인터페이스
 */
export interface ShopFilter {
  search?: string;
  type?: ShopType | ShopType[];
  services?: ServiceType | ServiceType[];
  rating?: number;
  distance?: number;
  location?: {
    latitude: number;
    longitude: number;
  };
  brands?: string[];
  open?: boolean;
  featured?: boolean;
  sortBy?: 'rating' | 'distance' | 'reviewCount' | 'name';
  sortOrder?: 'asc' | 'desc';
  page?: number;
  limit?: number;
} 

File: backend/schemas/shop.py
from pydantic import BaseModel, Field, EmailStr, HttpUrl
from typing import Optional, List, Dict
from datetime import datetime

# 정비소 생성 스키마
class ShopCreate(BaseModel):
    name: str = Field(..., max_length=100, description="정비소 이름")
    address: str = Field(..., max_length=200, description="정비소 주소")
    city: str = Field(..., max_length=50, description="도시")
    state: str = Field(..., max_length=50, description="주/도")
    postal_code: str = Field(..., max_length=20, description="우편번호")
    phone: str = Field(..., max_length=20, description="전화번호")
    email: Optional[EmailStr] = Field(None, description="이메일")
    website: Optional[str] = Field(None, max_length=100, description="웹사이트")
    description: Optional[str] = Field(None, description="정비소 설명")
    latitude: Optional[float] = Field(None, description="위도")
    longitude: Optional[float] = Field(None, description="경도")

# 정비소 업데이트 스키마
class ShopUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length=100, description="정비소 이름")
    address: Optional[str] = Field(None, max_length=200, description="정비소 주소")
    city: Optional[str] = Field(None, max_length=50, description="도시")
    state: Optional[str] = Field(None, max_length=50, description="주/도")
    postal_code: Optional[str] = Field(None, max_length=20, description="우편번호")
    phone: Optional[str] = Field(None, max_length=20, description="전화번호")
    email: Optional[EmailStr] = Field(None, description="이메일")
    website: Optional[str] = Field(None, max_length=100, description="웹사이트")
    description: Optional[str] = Field(None, description="정비소 설명")
    is_active: Optional[bool] = Field(None, description="활성 상태")
    latitude: Optional[float] = Field(None, description="위도")
    longitude: Optional[float] = Field(None, description="경도")

# 정비소 응답 스키마
class ShopResponse(BaseModel):
    id: int
    name: str
    address: str
    city: str
    state: str
    postal_code: str
    phone: str
    email: Optional[str] = None
    website: Optional[str] = None
    description: Optional[str] = None
    is_active: bool
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

# 정비소 목록 응답 스키마
class ShopListResponse(BaseModel):
    success: bool
    message: str
    count: int
    data: List[ShopResponse]
    
    class Config:
        from_attributes = True

# 정비소 상세 응답 스키마
class ShopDetailResponse(BaseModel):
    success: bool
    message: str
    data: ShopResponse
    
    class Config:
        from_attributes = True